File: scala/dao/implementations/FoodImpl.scala
package dao.implementations

import connection.DatabaseConnection
import dao.interfaces.FoodInterface
import entities.Food

import java.sql.SQLException

class FoodImpl extends FoodInterface{

  override def findFoodByName(name: String): Seq[Food] = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("SELECT * FROM food WHERE name = ?")

      stmt.setString(1, name)
      val res = stmt.executeQuery()
      var seq: Seq[Food] = Seq.empty[Food]
      while (res.next()) {
        val v = new Food(res.getLong("id"), res.getString("name"), res.getString("measurement"), res.getDouble("calories_multiplier"), res.getDouble("proteins_multiplier"),
          res.getDouble("fats_multiplier"), res.getDouble("carbohydrates_multiplier"), res.getDouble("fiber_multiplier"))
        seq = seq :+ v
      }
      seq
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        null
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }
  override def findById(id: Long): Option[Food] = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("SELECT * FROM food WHERE id = ?")
      stmt.setLong(1, id)
      val res = stmt.executeQuery()
      if (res.next()) {
        val v = new Food(res.getLong("id"), res.getString("name"), res.getString("measurement"), res.getDouble("calories_multiplier"), res.getDouble("proteins_multiplier"),
          res.getDouble("fats_multiplier"), res.getDouble("carbohydrates_multiplier"), res.getDouble("fiber_multiplier"))
        Option(v)
      } else{
        None
      }
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        None
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }

  override def addFood(v: Food): Boolean = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("INSERT INTO food VALUES (?, ?, ?, ?, ?, ?, ?, ?)")
      stmt.setLong(1, v.getId)
      stmt.setString(2, v.getName)
      stmt.setString(3, v.getMeasurement)
      stmt.setDouble(4, v.getCaloriesM)
      stmt.setDouble(5, v.getProteinsM)
      stmt.setDouble(6, v.getFatsM)
      stmt.setDouble(7, v.getCarbohydratesM)
      stmt.setDouble(8, v.getFiberM)
      stmt.executeQuery()
      true
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        false
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }
}

File: lab3/src/main/java/dao/implementations/FoodImpl.java
package dao.implementations;

import dao.interfaces.FoodInterface;
import entities.Food;
import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.EntityTransaction;
import jakarta.persistence.TypedQuery;

import java.util.List;
import java.util.Optional;

public class FoodImpl implements FoodInterface {
    private final EntityManagerFactory emf;

    public FoodImpl(EntityManagerFactory emf) {
        this.emf = emf;
    }

    @Override
    public List<Food> findFoodByName(String name) {
        EntityManager em = emf.createEntityManager();
        try {
            TypedQuery<Food> query = em.createQuery("SELECT f FROM Food f WHERE f.name = :name", Food.class);
            query.setParameter("name", name);
            return query.getResultList();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        } finally {
            em.close();
        }
    }

    @Override
    public Optional<Food> findById(Long id) {
        EntityManager em = emf.createEntityManager();
        try {
            return Optional.ofNullable(em.find(Food.class, id));
        } catch (Exception e) {
            e.printStackTrace();
            return Optional.empty();
        } finally {
            em.close();
        }
    }

    @Override
    public boolean addFood(Food food) {
        EntityManager em = emf.createEntityManager();
        EntityTransaction transaction = null;
        try {
            transaction = em.getTransaction();
            transaction.begin();
            em.persist(food);
            transaction.commit();
            return true;
        } catch (Exception e) {
            if (transaction != null) {
                transaction.rollback();
            }
            e.printStackTrace();
            return false;
        } finally {
            em.close();
        }
    }
}
